// paddleocr/dist/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
//#region src/constants.ts
const DEFAULT_DETECTION_OPTIONS = {
	padding: 0,
	mean: [
		.485 * 255,
		.456 * 255,
		.406 * 255
	],
	stdDeviation: [
		1 / .229 / 255,
		1 / .224 / 255,
		1 / .255 / 255
	],
	maxSideLength: 960,
	textPixelThreshold: .5,
	minimumAreaThreshold: 20,
	paddingBoxVertical: .4,
	paddingBoxHorizontal: .6,
	dilationKernelSize: 1
};
const DEFAULT_RECOGNITION_OPTIONS = {
	mean: [
		127.5,
		127.5,
		127.5
	],
	stdDeviation: [
		1 / 127.5,
		1 / 127.5,
		1 / 127.5
	],
	imageHeight: 48,
	charactersDictionary: []
};
const DEFAULT_RECOGNITION_ORDERING_OPTIONS = {
	sortByReadingOrder: true,
	sameLineThresholdRatio: .25
};
const DEFAULT_PROCESS_RECOGNITION_OPTIONS = { lineMergeThresholdRatio: .5 };
const DEFAULT_PADDLE_OPTIONS = {
	detection: DEFAULT_DETECTION_OPTIONS,
	recognition: DEFAULT_RECOGNITION_OPTIONS
};
//#endregion
//#region \0@oxc-project+runtime@0.115.0/helpers/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}
//#endregion
//#region \0@oxc-project+runtime@0.115.0/helpers/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}
//#endregion
//#region \0@oxc-project+runtime@0.115.0/helpers/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}
//#endregion
//#region \0@oxc-project+runtime@0.115.0/helpers/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}
//#endregion
//#region src/utils/image.ts
var Image = class Image {
	/**
	* 创建一个新的 Image 实例。
	* @param width 图像的宽度
	* @param height 图像的高度
	* @param data 图像数据，Uint8Array
	*/
	constructor(width, height, channels, data) {
		_defineProperty(this, "width", void 0);
		_defineProperty(this, "height", void 0);
		_defineProperty(this, "data", void 0);
		_defineProperty(this, "depth", void 0);
		_defineProperty(this, "channels", void 0);
		this.width = width;
		this.height = height;
		this.channels = channels;
		this.depth = 8;
		if (data) this.data = data;
		else {
			const length = width * height * 4;
			this.data = new Uint8Array(length);
		}
	}
	/**
	* 裁剪
	*/
	crop(options) {
		const { x, y, width, height } = options;
		if (x < 0 || y < 0 || x + width > this.width || y + height > this.height) throw new Error("Crop area is out of bounds");
		const croppedData = new Uint8Array(width * height * this.channels);
		for (let j = 0; j < height; j++) for (let i = 0; i < width; i++) {
			const srcIndex = ((y + j) * this.width + (x + i)) * this.channels;
			const dstIndex = (j * width + i) * this.channels;
			croppedData.set(this.data.subarray(srcIndex, srcIndex + this.channels), dstIndex);
		}
		return new Image(width, height, this.channels, croppedData);
	}
	/**
	* 将图片缩放到指定的尺寸w
	* @param options
	*/
	resize(options) {
		let { width, height } = options;
		if (width === void 0 && height === void 0) throw new Error("At least one of width or height must be specified");
		if (width === void 0) width = Math.round(this.width * ((height ?? this.height) / this.height));
		if (height === void 0) height = Math.round(this.height * (width / this.width));
		const srcW = this.width;
		const srcH = this.height;
		const dstW = width;
		const dstH = height;
		const channels = this.channels;
		const srcData = this.data;
		function triangle_kernel(x) {
			x = Math.abs(x);
			return x < 1 ? 1 - x : 0;
		}
		function clamp(v, min, max) {
			return Math.max(min, Math.min(max, v));
		}
		const tmpData = new Float32Array(srcW * dstH * channels);
		const ratioY = srcH / dstH;
		const sratioY = ratioY < 1 ? 1 : ratioY;
		const supportY = 1 * sratioY;
		for (let outy = 0; outy < dstH; outy++) {
			const inputy = (outy + .5) * ratioY - .5;
			const left = Math.max(0, Math.floor(inputy - supportY));
			const right = Math.min(srcH, Math.ceil(inputy + supportY));
			const ws = [];
			let sum = 0;
			for (let i = left; i < right; i++) {
				const w = triangle_kernel((i - inputy) / sratioY);
				ws.push(w);
				sum += w;
			}
			for (let i = 0; i < ws.length; i++) ws[i] /= sum;
			for (let x = 0; x < srcW; x++) for (let c = 0; c < channels; c++) {
				let t = 0;
				for (let i = 0; i < ws.length; i++) {
					const srcIdx = ((left + i) * srcW + x) * channels + c;
					t += srcData[srcIdx] * ws[i];
				}
				tmpData[(outy * srcW + x) * channels + c] = t;
			}
		}
		const dstData = new Uint8Array(dstW * dstH * channels);
		const ratioX = srcW / dstW;
		const sratioX = ratioX < 1 ? 1 : ratioX;
		const supportX = 1 * sratioX;
		for (let outx = 0; outx < dstW; outx++) {
			const inputx = (outx + .5) * ratioX - .5;
			const left = Math.max(0, Math.floor(inputx - supportX));
			const right = Math.min(srcW, Math.ceil(inputx + supportX));
			const ws = [];
			let sum = 0;
			for (let i = left; i < right; i++) {
				const w = triangle_kernel((i - inputx) / sratioX);
				ws.push(w);
				sum += w;
			}
			for (let i = 0; i < ws.length; i++) ws[i] /= sum;
			for (let y = 0; y < dstH; y++) for (let c = 0; c < channels; c++) {
				let t = 0;
				for (let i = 0; i < ws.length; i++) {
					const srcIdx = (y * srcW + (left + i)) * channels + c;
					t += tmpData[srcIdx] * ws[i];
				}
				dstData[(y * dstW + outx) * channels + c] = Math.round(clamp(t, 0, 255));
			}
		}
		return new Image(dstW, dstH, channels, dstData);
	}
	/**
	* 为图片添加指定颜色的边距，默认为透明的
	* @param options
	*/
	padding(options) {
		let { padding, vertical, horizontal, top, bottom, left, right, color } = options;
		if (typeof padding === "number") top = bottom = left = right = padding;
		else {
			if (typeof vertical === "number") top = bottom = vertical;
			if (typeof horizontal === "number") left = right = horizontal;
		}
		top = top ?? 0;
		bottom = bottom ?? 0;
		left = left ?? 0;
		right = right ?? 0;
		color = color ?? [
			0,
			0,
			0,
			0
		];
		const newW = this.width + left + right;
		const newH = this.height + top + bottom;
		const newData = new Uint8Array(newW * newH * 4);
		for (let y = 0; y < newH; y++) for (let x = 0; x < newW; x++) {
			const idx = (y * newW + x) * 4;
			newData[idx] = color[0];
			newData[idx + 1] = color[1];
			newData[idx + 2] = color[2];
			newData[idx + 3] = color[3];
		}
		for (let y = 0; y < this.height; y++) for (let x = 0; x < this.width; x++) {
			const srcIdx = (y * this.width + x) * 4;
			const dstIdx = ((y + top) * newW + (x + left)) * 4;
			newData.set(this.data.subarray(srcIdx, srcIdx + 4), dstIdx);
		}
		return new Image(newW, newH, this.channels, newData);
	}
	/**
	* 将当前图像转换为张量格式，以便输入到onnx模型
	* @param options
	*/
	tensor(options) {
		const mean = options.mean_values;
		const norm = options.norm_values;
		const width = this.width;
		const height = this.height;
		const numChannels = 3;
		const rgbaData = this.data;
		const tensor = new Float32Array(width * height * numChannels);
		for (let h = 0; h < height; h++) for (let w = 0; w < width; w++) {
			const pixelIndex = (h * width + w) * this.channels;
			const tensorIndex = h * width + w;
			for (let c = 0; c < numChannels; c++) {
				const normalizedValue = rgbaData[pixelIndex + c] * norm[c] - mean[c] * norm[c];
				tensor[c * height * width + tensorIndex] = normalizedValue;
			}
		}
		return tensor;
	}
	/**
	* 灰度图阈值方法，大于阈值的像素点设为255，小于等于阈值的设为0
	* @param options
	*/
	threshold(options) {
		const threshold = options.threshold ?? 128;
		const width = this.width;
		const height = this.height;
		const binData = new Uint8Array(width * height);
		for (let i = 0; i < width * height; i++) binData[i] = this.data[i * this.channels] > threshold ? 255 : 0;
		return new Image(width, height, 1, binData);
	}
	/**
	* 膨胀操作，使用指定的范数和核大小
	* 进行处理的图片像素是0和255，膨胀255的像素点
	* 返回一个新的图片
	* @param options
	*/
	dilate(options = {}) {
		const { norm = "LInf", k = 1 } = options;
		if (norm !== "LInf") throw new Error("Only LInf norm is supported");
		if (this.channels !== 1) throw new Error("Dilate only supports single channel (grayscale) images");
		const width = this.width;
		const height = this.height;
		const src = this.data;
		const INF = 999999;
		const dist = new Uint16Array(width * height);
		for (let i = 0; i < width * height; i++) dist[i] = src[i] > 0 ? 0 : INF;
		for (let y = 0; y < height; y++) for (let x = 0; x < width; x++) {
			const idx = y * width + x;
			if (dist[idx] === 0) continue;
			let minDist = INF;
			for (let dy = -1; dy <= 0; dy++) for (let dx = -1; dx <= 1; dx++) {
				if (dx === 0 && dy === 0) continue;
				const nx = x + dx, ny = y + dy;
				if (nx >= 0 && nx < width && ny >= 0 && ny < height) {
					const nidx = ny * width + nx;
					minDist = Math.min(minDist, dist[nidx] + 1);
				}
			}
			dist[idx] = Math.min(dist[idx], minDist);
		}
		for (let y = height - 1; y >= 0; y--) for (let x = width - 1; x >= 0; x--) {
			const idx = y * width + x;
			if (dist[idx] === 0) continue;
			let minDist = INF;
			for (let dy = 0; dy <= 1; dy++) for (let dx = -1; dx <= 1; dx++) {
				if (dx === 0 && dy === 0) continue;
				const nx = x + dx, ny = y + dy;
				if (nx >= 0 && nx < width && ny >= 0 && ny < height) {
					const nidx = ny * width + nx;
					minDist = Math.min(minDist, dist[nidx] + 1);
				}
			}
			dist[idx] = Math.min(dist[idx], minDist);
		}
		const out = new Uint8Array(width * height);
		for (let i = 0; i < width * height; i++) out[i] = dist[i] <= k ? 255 : 0;
		return new Image(width, height, 1, out);
	}
	/**
	* 获取图像中的轮廓
	* @returns
	*/
	contours(options = {}) {
		const minArea = options.minArea ?? 1;
		const width = this.width;
		const height = this.height;
		const bin = new Uint8Array(width * height);
		for (let i = 0; i < width * height; i++) bin[i] = this.data[i] > 0 ? 1 : 0;
		const visited = new Uint8Array(width * height);
		const boxes = [];
		const at = (x, y) => y * width + x;
		for (let y = 0; y < height; y++) for (let x = 0; x < width; x++) if (bin[at(x, y)] && !visited[at(x, y)]) {
			let minX = x, minY = y, maxX = x, maxY = y, area = 0;
			const queue = [[x, y]];
			visited[at(x, y)] = 1;
			while (queue.length) {
				const current = queue.shift();
				if (!current) break;
				const [cx, cy] = current;
				area++;
				minX = Math.min(minX, cx);
				minY = Math.min(minY, cy);
				maxX = Math.max(maxX, cx);
				maxY = Math.max(maxY, cy);
				for (const [dx, dy] of [
					[-1, 0],
					[1, 0],
					[0, -1],
					[0, 1],
					[-1, -1],
					[1, -1],
					[-1, 1],
					[1, 1]
				]) {
					const nx = cx + dx, ny = cy + dy;
					if (nx >= 0 && nx < width && ny >= 0 && ny < height && bin[at(nx, ny)] && !visited[at(nx, ny)]) {
						visited[at(nx, ny)] = 1;
						queue.push([nx, ny]);
					}
				}
			}
			if (area >= minArea) boxes.push({
				x: minX,
				y: minY,
				width: maxX - minX + 1,
				height: maxY - minY + 1
			});
		}
		return boxes;
	}
	/**
	* 在图像上绘制矩形，支持线宽
	* @param x 左上角x
	* @param y 左上角y
	* @param width 矩形宽度
	* @param height 矩形高度
	* @param color 颜色 [r,g,b,a]
	* @param lineWidth 线宽
	*/
	rect(options) {
		const { x, y, width, height, color = [], lineWidth = 1 } = options;
		if (!color.length) color.push(...Array(this.channels).fill(255));
		if (this.channels !== color.length) throw new Error(`Color length ${color.length} does not match image channels ${this.channels}`);
		for (let dy = 0; dy < lineWidth; dy++) for (let i = 0; i < width; i++) {
			const yy = y + dy;
			const xx = x + i;
			if (yy >= 0 && yy < this.height && xx >= 0 && xx < this.width) {
				const idx = (yy * this.width + xx) * this.channels;
				this.data.set(color, idx);
			}
			const by = y + height - 1 - dy;
			if (by >= 0 && by < this.height && xx >= 0 && xx < this.width) {
				const idx = (by * this.width + xx) * this.channels;
				this.data.set(color, idx);
			}
		}
		for (let dx = 0; dx < lineWidth; dx++) for (let j = 0; j < height; j++) {
			const xx = x + dx;
			const yy = y + j;
			if (xx >= 0 && xx < this.width && yy >= 0 && yy < this.height) {
				const idx = (yy * this.width + xx) * this.channels;
				this.data.set(color, idx);
			}
			const rx = x + width - 1 - dx;
			if (rx >= 0 && rx < this.width && yy >= 0 && yy < this.height) {
				const idx = (yy * this.width + rx) * this.channels;
				this.data.set(color, idx);
			}
		}
	}
};
//#endregion
//#region src/processor/detection.ts
/**
* Service for detecting text regions in images
*/
var DetectionService = class DetectionService {
	constructor(ortModule, session, options = {}) {
		_defineProperty(this, "options", void 0);
		_defineProperty(this, "session", void 0);
		_defineProperty(this, "ortModule", void 0);
		this.session = session;
		this.ortModule = ortModule;
		this.options = {
			...DEFAULT_DETECTION_OPTIONS,
			...options
		};
	}
	/**
	* Main method to run text detection on an image
	* @param image ArrayBuffer of the image or Canvas
	*/
	async run(image, options = {}) {
		const { onProgress, ...runtimeOverrides } = options;
		const runtimeOptions = this.resolveRuntimeOptions(runtimeOverrides);
		const input = await this.preprocessDetection(image, runtimeOptions);
		onProgress?.({
			type: "det",
			stage: "preprocess",
			progress: this.createProgress(1)
		});
		const detection = await this.runInference(input.tensor, input.resizeParams);
		onProgress?.({
			type: "det",
			stage: "infer",
			progress: this.createProgress(2)
		});
		const detectedBoxes = detection ? this.postprocessDetection(detection, input, runtimeOptions) : [];
		onProgress?.({
			type: "det",
			stage: "postprocess",
			progress: this.createProgress(3),
			detectedCount: detectedBoxes.length
		});
		return detectedBoxes;
	}
	resolveRuntimeOptions(options = {}) {
		return {
			...this.options,
			...options
		};
	}
	createProgress(current) {
		return {
			current,
			remain: DetectionService.TOTAL_PROGRESS_STEPS - current,
			total: DetectionService.TOTAL_PROGRESS_STEPS
		};
	}
	/**
	* Preprocess an image for text detection
	*/
	async preprocessDetection(image, runtimeOptions) {
		const resizeParams = this.calculateResizeDimensions(image, runtimeOptions);
		return {
			tensor: image.resize({
				width: resizeParams.dstWidth,
				height: resizeParams.dstHeight
			}).tensor({
				mean_values: runtimeOptions.mean,
				norm_values: runtimeOptions.stdDeviation
			}),
			resizeParams
		};
	}
	/**
	* Calculate dimensions for resizing the image
	*/
	calculateResizeDimensions(image, runtimeOptions) {
		const MAX_SIDE_LEN = runtimeOptions.maxSideLength;
		const { width: srcWidth, height: srcHeight } = image;
		const ratio = srcWidth > srcHeight ? MAX_SIDE_LEN / srcWidth : MAX_SIDE_LEN / srcHeight;
		let dstWidth = Math.floor(srcWidth * ratio);
		let dstHeight = Math.floor(srcHeight * ratio);
		if (dstWidth % 32 !== 0) dstWidth = Math.max(Math.floor(dstWidth / 32) * 32, 32);
		if (dstHeight % 32 !== 0) dstHeight = Math.max(Math.floor(dstHeight / 32) * 32, 32);
		const scaleWidth = dstWidth / srcWidth;
		const scaleHeight = dstHeight / srcHeight;
		return {
			srcHeight,
			srcWidth,
			dstHeight,
			dstWidth,
			scaleWidth,
			scaleHeight
		};
	}
	/**
	* Run the detection model inference
	*/
	async runInference(tensor, resizeParams) {
		const feeds = { x: new this.ortModule.Tensor("float32", tensor, [
			1,
			3,
			resizeParams.dstHeight,
			resizeParams.dstWidth
		]) };
		const outputTensor = (await this.session.run(feeds))[this.session.outputNames[0] || "fetch_name_0"];
		if (!outputTensor) return null;
		return outputTensor.data;
	}
	/**
	* Process detection results to extract bounding boxes
	*/
	postprocessDetection(detection, input, runtimeOptions) {
		const { dstWidth, dstHeight } = input.resizeParams;
		return new Image(dstWidth, dstHeight, 1, new Uint8Array(detection.map((v) => Math.round(v * 255)))).threshold({ threshold: 255 * runtimeOptions.textPixelThreshold }).dilate({
			norm: "LInf",
			k: runtimeOptions.dilationKernelSize
		}).contours({ minArea: runtimeOptions.minimumAreaThreshold }).map((box) => {
			const paddedBox = this.applyPaddingToRect(box, dstWidth, dstHeight, runtimeOptions);
			return this.convertToOriginalCoordinates(paddedBox, input.resizeParams);
		});
	}
	/**
	* Apply padding to a rectangle
	*/
	applyPaddingToRect(rect, maxWidth, maxHeight, runtimeOptions) {
		const paddingVertical = runtimeOptions.paddingBoxVertical;
		const paddingHorizontal = runtimeOptions.paddingBoxHorizontal;
		const verticalPadding = Math.round(rect.height * paddingVertical);
		const horizontalPadding = Math.round(rect.height * paddingHorizontal);
		let x = rect.x - horizontalPadding;
		let y = rect.y - verticalPadding;
		let width = rect.width + 2 * horizontalPadding;
		let height = rect.height + 2 * verticalPadding;
		x = Math.max(0, x);
		y = Math.max(0, y);
		const rightEdge = Math.min(maxWidth, rect.x + rect.width + horizontalPadding);
		const bottomEdge = Math.min(maxHeight, rect.y + rect.height + verticalPadding);
		width = rightEdge - x;
		height = bottomEdge - y;
		return {
			x,
			y,
			width,
			height
		};
	}
	/**
	* Convert coordinates from resized image back to original image
	*/
	convertToOriginalCoordinates(rect, resizeParams) {
		const scaledX = rect.x / resizeParams.scaleWidth;
		const scaledY = rect.y / resizeParams.scaleHeight;
		const scaledWidth = rect.width / resizeParams.scaleWidth;
		const scaledHeight = rect.height / resizeParams.scaleHeight;
		const x = Math.max(0, Math.round(scaledX));
		const y = Math.max(0, Math.round(scaledY));
		return {
			x,
			y,
			width: Math.min(resizeParams.srcWidth - x, Math.round(scaledWidth)),
			height: Math.min(resizeParams.srcHeight - y, Math.round(scaledHeight))
		};
	}
};
_defineProperty(DetectionService, "TOTAL_PROGRESS_STEPS", 3);
//#endregion
//#region src/processor/recognition.ts
/**
* Service for detecting and recognizing text in images
*/
var RecognitionService = class {
	constructor(ortModule, session, options = {}) {
		_defineProperty(this, "options", void 0);
		_defineProperty(this, "session", void 0);
		_defineProperty(this, "ortModule", void 0);
		this.session = session;
		this.ortModule = ortModule;
		this.options = {
			...DEFAULT_RECOGNITION_OPTIONS,
			...options
		};
	}
	/**
	* Main method to run text recognition on an image with detected regions
	* @param image The original image buffer or image in Canvas
	* @param detection Array of bounding boxes from text detection
	* @returns Array of recognition results with text and bounding box, sorted in reading order
	*/
	async run(image, detection, options) {
		const recognitionOptions = this.resolveRuntimeOptions(options?.recognition);
		const orderingOptions = this.resolveOrderingOptions(options?.ordering);
		const validBoxes = this.sortBoxesByReadingOrder(detection.filter((box) => box.width > 0 && box.height > 0), orderingOptions);
		const results = [];
		const charWhiteListSet = options?.charWhiteList?.length ? new Set(options.charWhiteList) : void 0;
		const total = validBoxes.length;
		const onProgress = options?.onProgress;
		onProgress?.({
			type: "rec",
			stage: "start",
			progress: this.createProgress(0, total)
		});
		for (const [i, box] of validBoxes.entries()) {
			const result = await this.processBox({
				image,
				index: i,
				box,
				charWhiteSet: charWhiteListSet
			}, recognitionOptions);
			if (result) results.push(result);
			onProgress?.({
				type: "rec",
				stage: "item",
				progress: this.createProgress(i + 1, total),
				index: i,
				box,
				result: result ?? void 0
			});
		}
		onProgress?.({
			type: "rec",
			stage: "complete",
			progress: this.createProgress(total, total)
		});
		return results;
	}
	resolveRuntimeOptions(options = {}) {
		return {
			...this.options,
			...options
		};
	}
	resolveOrderingOptions(options = {}) {
		return {
			...DEFAULT_RECOGNITION_ORDERING_OPTIONS,
			...options
		};
	}
	/**
	* Process a single text box
	*/
	async processBox(task, runtimeOptions) {
		const { image, box } = task;
		const resizedCrop = image.crop(box).resize({ height: runtimeOptions.imageHeight });
		const tensor = resizedCrop.tensor({
			mean_values: runtimeOptions.mean,
			norm_values: runtimeOptions.stdDeviation
		});
		const inputTensor = new this.ortModule.Tensor("float32", tensor, [
			1,
			3,
			resizedCrop.height,
			resizedCrop.width
		]);
		const { data: outputData, dims: shape } = await this.runInference(inputTensor);
		const [, sequenceLength, numClasses] = shape;
		const { text: recognizedText, confidence } = this.ctcLabelDecode(outputData, sequenceLength, numClasses, runtimeOptions, task.charWhiteSet);
		return {
			text: recognizedText,
			box,
			confidence
		};
	}
	/**
	* Sort recognition results by reading order (top to bottom, left to right)
	*/
	sortBoxesByReadingOrder(boxes, orderingOptions) {
		if (!orderingOptions.sortByReadingOrder) return [...boxes];
		return [...boxes].sort((boxA, boxB) => {
			if (Math.abs(boxA.y - boxB.y) < (boxA.height + boxB.height) * orderingOptions.sameLineThresholdRatio) return boxA.x - boxB.x;
			return boxA.y - boxB.y;
		});
	}
	createProgress(current, total) {
		return {
			current,
			remain: total - current,
			total
		};
	}
	/**
	* Runs the ONNX inference session with the prepared tensor
	*/
	async runInference(inputTensor) {
		const feeds = { x: inputTensor };
		const results = await this.session.run(feeds);
		const outputNodeName = Object.keys(results)[0];
		const outputTensor = results[outputNodeName];
		if (!outputTensor) throw new Error(`Recognition output tensor '${outputNodeName}' not found. Available keys: ${Object.keys(results)}`);
		return outputTensor;
	}
	ctcLabelDecode(logits, sequenceLength, numClasses, runtimeOptions, charWhiteSet) {
		const dict = runtimeOptions.charactersDictionary;
		let text = "";
		const scores = [];
		let lastIndex = -1;
		for (let t = 0; t < sequenceLength; t++) {
			let maxScore = 0;
			let maxScoreIndex = 0;
			const offset = t * numClasses;
			for (let i = 0; i < numClasses; i++) {
				const val = logits[offset + i];
				if (val > maxScore) {
					maxScore = val;
					maxScoreIndex = i;
				}
			}
			if (maxScoreIndex === lastIndex) continue;
			lastIndex = maxScoreIndex;
			if (maxScoreIndex === 0) continue;
			const char = dict[maxScoreIndex] || "";
			if (charWhiteSet && !charWhiteSet.has(char) && char !== " ") continue;
			text += char;
			scores.push(maxScore);
		}
		return {
			text,
			confidence: scores.length > 0 ? scores.reduce((sum, score) => sum + score, 0) / scores.length : 0
		};
	}
};
//#endregion
//#region src/processor/paddle-ocr.ts
/**
* PaddleOcrService - Provides OCR functionality using PaddleOCR models
*
* This service can be used either as a singleton or as separate instances
* depending on your application needs.
*/
var PaddleOcrService = class PaddleOcrService {
	/**
	* Create a new PaddleOcrService instance
	* @param options Optional configuration options
	*/
	constructor(options) {
		_defineProperty(this, "options", void 0);
		_defineProperty(this, "detectionSession", null);
		_defineProperty(this, "detectionService", null);
		_defineProperty(this, "recognitionSession", null);
		_defineProperty(this, "recognitionService", null);
		if (!options?.ort) throw new Error("PaddleOcrService requires the 'ort' option to be set with onnxruntime-node or onnxruntime-wen.");
		this.options = {
			...DEFAULT_PADDLE_OPTIONS,
			...options || {}
		};
	}
	/**
	* Initialize the OCR service by loading models
	*/
	async initialize() {
		const ort = this.options.ort;
		if (!ort) throw new Error("PaddleOcrService requires the 'ort' option to be set with onnxruntime-node or onnxruntime-wen.");
		const detectionModelBuffer = this.options.detection?.modelBuffer;
		if (!detectionModelBuffer) throw new Error("Detection model buffer is required. Please provide a valid ONNX model.");
		this.detectionSession = await ort.InferenceSession.create(detectionModelBuffer);
		const { modelBuffer: _detectionModelBuffer, ...detectionOptions } = this.options.detection ?? {};
		this.detectionService = new DetectionService(ort, this.detectionSession, detectionOptions);
		const recognitionModelBuffer = this.options.recognition?.modelBuffer;
		if (!recognitionModelBuffer) throw new Error("Recognition model buffer is required. Please provide a valid ONNX model.");
		this.recognitionSession = await ort.InferenceSession.create(recognitionModelBuffer);
		const { modelBuffer: _recognitionModelBuffer, ...recognitionOptions } = this.options.recognition ?? {};
		this.recognitionService = new RecognitionService(ort, this.recognitionSession, recognitionOptions);
	}
	/**
	* Check if the service is initialized with models loaded
	*/
	isInitialized() {
		return this.detectionSession !== null && this.recognitionSession !== null;
	}
	/**
	* Create a new instance instead of using the singleton
	* This is useful when you need multiple instances with different models
	* @param options Configuration options for this specific instance
	*/
	static async createInstance(options) {
		const instance = new PaddleOcrService(options);
		await instance.initialize();
		return instance;
	}
	resolveDetectionRuntimeOptions(options = {}) {
		const { modelBuffer: _modelBuffer, ...instanceOptions } = this.options.detection ?? {};
		return {
			...DEFAULT_DETECTION_OPTIONS,
			...instanceOptions,
			...options
		};
	}
	resolveRecognitionRuntimeOptions(options = {}) {
		const { modelBuffer: _modelBuffer, ...instanceOptions } = this.options.recognition ?? {};
		return {
			...DEFAULT_RECOGNITION_OPTIONS,
			...instanceOptions,
			...options
		};
	}
	resolveRecognitionOrderingOptions(options = {}) {
		return {
			...DEFAULT_RECOGNITION_ORDERING_OPTIONS,
			...options
		};
	}
	/**
	* Runs object detection on the provided image input, then performs
	* recognition on the detected regions.
	*
	* @param image - The raw image data as an ArrayBuffer or Canvas.
	* @param options - Optional configuration for the recognition output, e.g., `{ flatten: true }`.
	* @return A promise that resolves to the OCR result, either grouped by lines or as a flat list.
	*/
	async recognize(input, options) {
		if (!this.detectionService || !this.recognitionService) throw new Error("PaddleOcrService is not initialized. Please call initialize() first.");
		const channels = input.data.length / (input.width * input.height);
		if (!Number.isInteger(channels) || channels < 1 || channels > 4) throw new Error(`Invalid input data: ${input.data} for image size ${input.width}x${input.height}. Expected 1, 3, or 4 channels.`);
		const detectionRuntimeOptions = this.resolveDetectionRuntimeOptions(options?.detection);
		const recognitionRuntimeOptions = this.resolveRecognitionRuntimeOptions(options?.recognition);
		const orderingOptions = this.resolveRecognitionOrderingOptions(options?.ordering);
		if (!recognitionRuntimeOptions.charactersDictionary?.length) throw new Error("Recognition charactersDictionary is required. Provide it in createInstance({ recognition }) or recognize(_, { recognition }).");
		let image = new Image(input.width, input.height, channels, input.data);
		const padding = detectionRuntimeOptions.padding;
		if (padding) image = image.padding({
			padding,
			color: [
				255,
				255,
				255,
				255
			]
		});
		const detection = await this.detectionService.run(image, {
			...detectionRuntimeOptions,
			onProgress: options?.onProgress
		});
		return await this.recognitionService.run(image, detection, {
			...options,
			detection: detectionRuntimeOptions,
			recognition: recognitionRuntimeOptions,
			ordering: orderingOptions
		});
	}
	/**
	* Processes raw recognition results to generate the final text,
	* grouped lines, and overall confidence.
	*/
	processRecognition(recognition, options) {
		const result = {
			text: "",
			lines: [],
			confidence: 0
		};
		if (!recognition.length) return result;
		result.confidence = recognition.reduce((sum, r) => sum + r.confidence, 0) / recognition.length;
		const processOptions = {
			...DEFAULT_PROCESS_RECOGNITION_OPTIONS,
			...options
		};
		let currentLine = [recognition[0]];
		let fullText = recognition[0].text;
		let avgHeight = recognition[0].box.height;
		for (let i = 1; i < recognition.length; i++) {
			const current = recognition[i];
			const previous = recognition[i - 1];
			if (Math.abs(current.box.y - previous.box.y) <= avgHeight * processOptions.lineMergeThresholdRatio) {
				currentLine.push(current);
				fullText += ` ${current.text}`;
				avgHeight = currentLine.reduce((sum, r) => sum + r.box.height, 0) / currentLine.length;
			} else {
				result.lines.push([...currentLine]);
				fullText += `\n${current.text}`;
				currentLine = [current];
				avgHeight = current.box.height;
			}
		}
		if (currentLine.length > 0) result.lines.push([...currentLine]);
		result.text = fullText;
		return result;
	}
	/**
	* Releases the onnx runtime session for both
	* detection and recognition model.
	*/
	async destroy() {
		await this.detectionSession?.release();
		await this.recognitionSession?.release();
	}
};
//#endregion
exports.DEFAULT_DETECTION_OPTIONS = DEFAULT_DETECTION_OPTIONS;
exports.DEFAULT_PADDLE_OPTIONS = DEFAULT_PADDLE_OPTIONS;
exports.DEFAULT_PROCESS_RECOGNITION_OPTIONS = DEFAULT_PROCESS_RECOGNITION_OPTIONS;
exports.DEFAULT_RECOGNITION_OPTIONS = DEFAULT_RECOGNITION_OPTIONS;
exports.DEFAULT_RECOGNITION_ORDERING_OPTIONS = DEFAULT_RECOGNITION_ORDERING_OPTIONS;
exports.DetectionService = DetectionService;
exports.PaddleOcrService = PaddleOcrService;
exports.RecognitionService = RecognitionService;
